### ISSUE TITLE
MD5 - Linux compatibility

### ISSUE THREAD

**Issue by Janiczek** [joshuaclayton/unused#49]
https://github.com/joshuaclayton/unused/blob/master/src/Unused/Cache/DirectoryFingerprint.hs

Linux (or, at least Ubuntu) doesn't have `md5` utility - it has `md5sum`. Maybe it would be a good idea to either
a) calculate MD5 in Haskell, or
b) find out which utility to use?

```
$ unused
Unused: analyzing 27269 terms

Calculating cache fingerprint... 
find: ‘md5’: No such file or directory
find: ‘md5’: No such file or directory
find: ‘md5’: No such file or directory
find: ‘md5’: No such file or directory
unused: readCreateProcess: find "." "-type" "f" "-not" "-path" "*/.git/*" "-not" "-name" "*/.buildpath" "-not" "-path" "*/.buildpath/*" "-not" "-name" "*/.project" "-not" "-path" "*/.project/*" "-not" "-name" "*/.settings" "-not" "-path" "*/.settings/*" "-not" "-name" "*/.idea" "-not" "-path" "*/.idea/*" "-not" "-path" "*/nbproject/*" "-not" "-name" "*/node_modules" "-not" "-path" "*/node_modules/*" "-not" "-name" "*/bower_components" "-not" "-path" "*/bower_components/*" "-not" "-name" "*/app/mobile/assets/bower_components" "-not" "-path" "*/app/mobile/assets/bower_components/*" "-not" "-path" "*/www/resources/css/**/*.min.css" "-not" "-path" "*/www/resources/css/both/*" "-not" "-path" "*/www/resources/css/ps/*" "-not" "-path" "*/www/resources/css/hs/*" "-not" "-path" "*/www/resources/js/*.min.js" "-not" "-path" "*/vendor/*" "-not" "-name" "*/composer.lock" "-not" "-path" "*/composer.lock/*" "-not" "-name" "*/checkstyle.xml" "-not" "-path" "*/checkstyle.xml/*" "-not" "-name" "*/composer.phar" "-not" "-path" "*/composer.phar/*" "-not" "-name" "*/phpmd.xml" "-not" "-path" "*/phpmd.xml/*" "-not" "-name" "*/www/assets/css/styles.css" "-not" "-path" "*/www/assets/css/styles.css/*" "-not" "-name" "*/www/assets/css/styles.min.css" "-not" "-path" "*/www/assets/css/styles.min.css/*" "-not" "-name" "*/.DS_Store" "-not" "-path" "*/.DS_Store/*" "-exec" "md5" "{}" "+" (exit 1): failed
```

**Comment by joshuaclayton** [comment]
@Janiczek ah, thanks for bringing this up. I definitely made a handful of initial assumptions to get this working.

I'm relatively new to Haskell - apart from shelling out `which md5` and `which md5sum` to see which returns a `0` exit code, do you know if there's a good way to check this?

As for performing this check within the application itself, I'm not necessarily opposed to that, but I'd be concerned that I'd introduce other performance implications or introduce something that's further non-portable. Standing on the shoulders of giants and all that 😄.

At any rate, I'll do a bit of digging and see what I find. Thanks for reporting this!

**Comment by Janiczek** [comment]
I didn't find a Haskell library that would take a directory and calculate its MD5 hash. The shelling out seems the best idea.

Ubuntu:
```
Prelude System.Process> readProcessWithExitCode "which" ["md5"] []
(ExitFailure 1,"","")
Prelude System.Process> readProcessWithExitCode "which" ["md5sum"] []
(ExitSuccess,"/usr/bin/md5sum\n","")
```

**Comment by Janiczek** [comment]
Also, for compatibility between `md5` and `md5sum` outputs, I found [this](https://www.garron.me/en/bits/how-to-md5sum-mac-os-x.html) for the OSX `md5` tool:

> If you need the same output format that md5sum has, use this.
> 
> `md5 -r [file.ext]`

**Comment by joshuaclayton** [comment]
@Janiczek I've opened https://github.com/joshuaclayton/unused/pull/51, which begins to address this. I did find https://hackage.haskell.org/package/directory-1.2.6.3/docs/System-Directory.html#v:findExecutables, which looks to hide having to shell out to `which`. I also have the error bubble up, halting the program, instead of silently skipping the caching aspects. This seems to be better than what was there previously, and if users on any other platforms where this doesn't attempt to run, they'll at least have more insight into where it's failing.

For further code-specific discussion, mind throwing it on that thread instead?

**Comment by joshuaclayton** [comment]
@Janiczek re: output compat between the two, I'll play around with a VM and see how things work out - specifically, I'm not sure which set of behavior I want to mirror because I'm unfamiliar with the output format of `md5sum`. For `md5`, the program is effectively piping the sorted output of `find`, which roughly translates to `find . -type f -not -path '*/.git/*' -exec md5 {} + | sort -k2 | md5`. The only output I see from that command is the SHA itself; if the output is different on your OS, please let me know and I can adjust accordingly.

**Comment by Janiczek** [comment]
```
$ md5sum readme.md 
397b892208ce4ba18f6ff16737da9e47  readme.md
```

I suspect `md5 -r` on your OSX box will be similar.

It doesn't allow for getting MD5 of a directory if you need that, though. There are workarounds: [1](http://unix.stackexchange.com/questions/35832/how-do-i-get-the-md5-sum-of-a-directorys-contents-as-one-sum), [2](http://stackoverflow.com/questions/1657232/how-can-i-calculate-an-md5-checksum-of-a-directory)

**Comment by joshuaclayton** [comment]
@Janiczek thanks for sharing! I'd zeroed in on roughly option 2 (http://stackoverflow.com/a/1658554) in my implementation. The difference between OS X and Linux at this point (also outlined in SO) is the result is the digest, two spaces, and a dash (which seems to normally represent a filename, but because it's receiving data piped from STDIN, the `-` represents lack of an actual file).

I introduced a commit https://github.com/joshuaclayton/unused/pull/51/commits/02989c60972db32be45414b0355df08a00f5bf22 to address that, ensuring consistency by grabbing all alphanumeric characters until the first non-alphanum. This removes the trailing whitespace and dash, so it should be consistent across platforms (I don't think it really matters, but the digest is used to generate the filenames and I didn't like the whitespace being present).

I'm also working on a Docker set up for this, primarily for more integration testing on non-OSX platforms. I was able to confirm that, after this fix, caching does work correctly and results are displayed as expected (hooray!). I'll try to snag further feedback on the actual implementation, and close this once it's merged. I should be able to get 0.5 pushed out in the coming days which will include this fix.
